### PR TITLE
Add artistImage in artist list view

### PR DESCRIPTION
**Pull request by rohitgeddam**
use subsonic api to fetch the artist detail  

artist_list_view

***This is just a draft to get work started step by step***

#847

### CODE REVIEW

**Comment by rohitgeddam** [comment]
***I would require the mbzArtistId to pull more information about the artist but it is empty.***

artist_api

***I tried to get artist information from getArtistInfo2 subsonic endpoint but it dosen't return musizbrainzId as expected from the subsonic docs [http://subsonic.org/pages/inc/api/examples/artistInfo2_example_1.xml](url),
I suspect this is happening because the subsonic api cant get the artist also the artistImage is a generic placeholder star image***

getArtistInfo2

If we could just get the mbzArtistId we can just pull it from musicbrainz api
Is there any other way to get artist info?

**Comment by deluan** [comment]
Did you configured External Integrations? https://www.navidrome.org/docs/usage/external_integrations/

**Comment by deluan** [comment]
> I tried to get artist information from getArtistInfo2 subsonic endpoint but it dosen't return musizbrainzId as expected from the subsonic docs

The Subsonic docs says it is an optional field. From the [API XSD](http://subsonic.org/pages/inc/api/schema/subsonic-rest-api-1.16.1.xsd):
```xml
<xs:complexType name="ArtistInfoBase">
<xs:sequence>
<xs:element name="biography" type="xs:string" minOccurs="0" maxOccurs="1"/>
<xs:element name="musicBrainzId" type="xs:string" minOccurs="0" maxOccurs="1"/>
<xs:element name="lastFmUrl" type="xs:string" minOccurs="0" maxOccurs="1"/>
<xs:element name="smallImageUrl" type="xs:string" minOccurs="0" maxOccurs="1"/>
<xs:element name="mediumImageUrl" type="xs:string" minOccurs="0" maxOccurs="1"/>
<xs:element name="largeImageUrl" type="xs:string" minOccurs="0" maxOccurs="1"/>
</xs:sequence>
</xs:complexType>
```

It is not being returned because it is unavailable (see the previous comment to set it up).

> If we could just get the mbzArtistId we can just pull artist info from musicbrainz api

Again, we don't want the UI making calls to external services (unless there's no better way). And this is already handled in the server. Please search for `lastfm` in the codebase, or follow the `GetSartistInfo2` implementation.

**Comment by rohitgeddam** [comment]
> Did you configured External Integrations? https://www.navidrome.org/docs/usage/external_integrations/

I completely overlooked it. I should read the documentation more carefully, could have saved many hours?

**Comment by deluan** [comment]
Hey @rohitgeddam, are you still planning on finishing this?

**Comment by rohitgeddam** [comment]
Yes, i will resume my work today, been busy with my exams.

**Comment by rohitgeddam** [comment]
## Rows
    [ ] multi select boxes
    [x] Artist with image
    [ ] Country ( either from tags or from last.fm/musizbrainz) require(musizBrainz)
    [ ] Years (Year range of artists songs eg 1984-2013,)   require(musicBrainz)
    [ ] Genres (list all generes ranked by number of songs )
    [x] songs/albums ( number of songs/albums eg 48/5)
    [ ] Total Time ( total duration of all music by artist eg 245:12)
    [x] plays ( cumulative plays for all tracks)
    [ ] last played 
    [ ] popularity ( last.fm has a metric for that? google trends?) NR
    [ ] rating ( avg rating 1-5)

## Filters
    [ ] only album artists
    [ ] genre
    [ ] last played
    [ ] decade
    [ ] role

**Comment by rohitgeddam** [comment]
artist_image

Artist images are getting populated only for the artists for which I called `getArtistInfo2` API route, using postman,
I think as a workaround I can call the `getArtistInfo2` API route ***only for artists whose imageurls are empty,*** in react `useEffect` hook,
that way we can populate images for all artists

@deluan as you already mentioned it might take some time to populate the imageurl the first time, another way as you mentioned is to populate these images in the background using a worker pool.

Since populating these fields is only a one-time thing, we can go ahead with the `useEffect` way, for all subsequent requests we will already have these fields populated.

@deluan what do you think?

**Comment by rohitgeddam** [comment]
artist_view_image_populate

After adding the worker, artist images are getting populated correctly, we can also extend this worker to fetch more details about the artist using other apis.

File: cmd/root.go [hunk 1]
```diff
@@ -55,12 +55,34 @@ func runNavidrome() {
...
+	g.Add(startWorker())
 
 	if err := g.Run(); err != nil {
 		log.Error("Fatal error in Navidrome. Aborting", err)
 	}
 }
 
+func startWorker() (func() error, func(err error)) {
+	ctx, cancel := context.WithCancel(context.Background())
+	interval := time.Minute
```
Comment (rohitgeddam) [inline cmd/root.go:67]: value of `interval` will be changed to be read from the config file

File: cmd/worker.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+func populateArtistDetails(ctx context.Context, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	ds := persistence.New()
+	ArtistRepo := ds.Artist(ctx)
+	es := core.NewExternalMetadata(ds)
+	for {
+		fmt.Println("Worker IS RUNNING IN INFI LOOP")
```
Comment (rohitgeddam) [inline cmd/worker.go:19]: forgot to remove this line!

**Comment by deluan** [comment]
Hey, thanks for taking a stab at this. Some considerations:
- I think the scope of this should be focused on showing artist images, the other information (country, years, etc...) requires more upfront architectural thinking
- When I talked about a worker pool, I was referring to the code we already have in `utils/pools.go` and not about introducing a new agent as you did. Although the solution you implemented works (obviously :) ), it does not make sense to keep updating the artist info continuously. We only need to do it at import time, when a new artist is added to the DB. This would behave similar to what we already do to pre-cache album coverart.

Let me know if that makes sense.

**Comment by rohitgeddam** [comment]
Screenshot from 2021-04-04 21-54-43

In the above ss, Is that a good place to `submit` work to worker pool, since this code is run everytime a new artist is inserted

**Comment by deluan** [comment]
I don't think it should be the persistence layer's responsibility to update artist info from external sources. It should be done in the scanner during the import process. As I said, album cover art pre-cacheing is already done this way (take a look at the `CacheWarmer`).

I suggest you break this into two parts:
1) Display the artist image in the UI. If empty call `getArtistInfo2`
2) Automatically call the updateArtistInfo when importing new/updated artists in the backend

**Comment by certuna** [comment]
I think in the list view we should have smaller images like this: https://marmelab.com/react-admin-demo/#/customers and save the big images for a grid view.

**Comment by rohitgeddam** [comment]
artist_view

File: scanner/tag_scanner.go [hunk 3]
```diff
@@ -326,9 +327,30 @@ func (s *TagScanner) deleteOrphanSongs(ctx context.Context, dir string, tracksTo
...
+type artistItem struct {
+	artistId string
+}
+
+func (s *TagScanner) execute(workload interface{}) {
+	es := core.NewExternalMetadata(s.ds)
+	ctx := context.Background()
+	item := workload.(artistItem)
+	_, err := es.UpdateArtistInfo(ctx, item.artistId, 0, false)
+	if err != nil {
```
Comment (rohitgeddam) [inline scanner/tag_scanner.go:339]: I added a checkpoint here to investigate the code, but the debugger doesn't halt here.
`WARN[0002]: data not found`
artist_view_error

Comment (rohitgeddam) [reply scanner/tag_scanner.go:339]: I am investigating on this

**Comment by pkspyder007** [comment]
> The density of information is pretty low: when I'm browsing an artists' list, I want to see a maximum of artists, here most of the space is empty, to accommodate the pictures' presence.

@jvoisin  @rohitgeddam  We can reduce the image size to a avatar for more rows to be viewed at once. 
And we can show a bigger image on hover.

**Comment by deluan** [comment]
Hey @rohitgeddam , will you be able to finish this? Please let me know

**Comment by deluan** [comment]
Hey, I'm closing this as I'll need to rewrite the UI with the new React-Admin version (v4) and will try to implement this in a different way. Thanks anyway for the contribution!

**Pull request closed without merging**